pycc/local.py

```diff
@@ -366,6 +360,277 @@ def _build_LPNO(self):
+                r2[i,j] = r2_ij + r2_ij.swapaxes(0,1).swapaxes(2,3)
+    
+            # probably have to update D
+            t2 += r2/Dijab
+            emp2 = contract('ab,ab->', t2_ij, L_ij[o,o,v,v])
+
+            rmsd = np.sqrt(contract('ijab,ijab->', r2/Dijab, r2/Dijab))
+
+            emp2 = contract('ijab,ijab->', t2, self.H.L[o,o,v,v])
+            ediff = emp2 - elast
+
+            print("MP2 Iter %3d: MP2 Ecorr = %.15f  dE = % .5E  rmsd = % .5E" % (niter, emp2, ediff, rmsd))
+    
+    def _local_MP2_loop(self,local,S):
+        """
+        Performs the MP2 loop in the transformed local basis (PAO, PNO)
+
+        Attributes
+        ----------
+        Q: transform between canonical VMO and local spaces
+        L: transform between local and semicanonical local spaces
+        eps: semicananonical local energies
+
+        Notes
+        -----
+        - References: Hampel & Werner 1996 [10.1063/1.471289], Pinski & riplinger & Valeev [10.1063/1.4926879], 
+          and Zach's DLPNO-MP2 python code
+
+        """
+        Q = self.Q
+        L = self.L
+        o = slice(0,self.no)
+        v = slice(self.no, self.no + self.nv)
+
+        # noting the occupied fock matrix
+        F_occ = self.H.F[o,o]
+        
+        # initial guess amplitude in the local MO basis
+        eps_occ = np.diag(self.H.F)[o]
+        eps_vir = np.diag(self.H.F)[v]
+        Dijab = eps_occ.reshape(-1,1,1,1) + eps_occ.reshape(-1,1,1) - eps_vir.reshape(-1,1) - eps_vir
+        t2 = self.H.ERI[o,o,v,v]/Dijab
+        
+        # Transformation basis for the virtual-virtual blocks of the Fock matrix, ERI tensor,
+        # T2 tensor, L tensor, and overlap matrix to PAO/PNO basis for each pair
+        
+        # removed the PAO code for now, similar code to PNO with the addition of overlap terms due to nonorthogonal 
+        # virtual basis
+
+        if local.upper() in ["LPNO"]:
+       
+        
+        # C_ij = ao by pno_o       
+        # ERI_ij = pno_o by pno_o
+        # t2_ij = pno_o by pno_o 
+
+            C_ij = []  
+            ERI_ij = []
+            t2_ij = []
+   
+            # Transformed virtual-virtual block of the overlap, Fock, ERI, t2 in pair domains through modification 
+            # of the atomic orbital coefficients
+            # All the transformation leads to the semicanonical pno basis 
+            
+            for ij in range(self.no*self.no):
+                i = ij // self.no
+                j = ij % self.no
+            
+ 
+                # C_ij = ao by pno_n
+                # C = ao by lmo
+                # Q = lmo by pno_r
+                
+                C_ij.append(self.C[:,v] @ Q[ij])
+
+                # C_ij = ao by pno_o
+                # L = pno_r by pno_o
+                
+                C_ij[ij] = C_ij[ij] @ L[ij]
+                
+                # S_ij = lmo by pno_o
+                # S = ao by ao 
+                S_ij = self.C.T @ S @ C_ij[ij]                  
```

Comment:
since the left-hand C is "full C" (nao x nmo), `C.T @ S` returns a matrix of size (nmo x nao). if you want only the "virtual" component of this, I think you should take the virtual block of the resultant matrix (now the last rows, rather than columns). eg:

```
tmp = self.C.T @ S
S_ij = tmp[v,:] @ C_ij[ij]
```

this should result in a transformed "overlap" for only the virtual orbitals, and it should be (nPNO x nPNO). currently, this returns a matrix of size (nMO x nPNO).